File: Top_Interview/98_ValidateBinarySearchTree/98.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {
    public boolean helper(TreeNode root, Integer lower, Integer upper){
        if(root==null){
            return true;
        }
        
        int val = root.val;
        if(lower!=null && val<=lower){
            return false;
        }
        if(upper!=null && val>=upper){
            return false;
        }
        
        if(!helper(root.right, val, upper)){
            return false;
        }
        if(!helper(root.left, lower, val)){
            return false;
        }
        
        return true;
    }
    public boolean isValidBST(TreeNode root) {
        return helper(root, null, null);
    }
}

File: Top_Interview/98_ValidateBinarySearchTree/98.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    bool helper(TreeNode* root, int minval, int maxval){
        if(root==NULL){
            return true;
        }
        if(root->val<=minval || root->val>=maxval){
            return false;
        }
        return helper(root->left, minval, root->val)
            && helper(root->right, root->val, maxval);
    }
    bool isValidBST(TreeNode* root) {
        return helper(root, INT_MIN, INT_MAX);
    }
};
